File: ps2/template/Exercises2-5-Matlab/Ex3.m

%% I. Exercise 3: Bewley model
% Specific parameter adjustments
mpar.mink = 0;
mpar.maxk = 3;    % Maximimum assets
par.b     = mpar.mink;
disp('-------------------------')
disp('Exercise 3: Bewley model')
disp('-------------------------')
disp('Economic parameters')
disp(par) % Display economic parameters
%% 1. Generate grids, Meshes and Income
gri.k   = exp(linspace(0,log(mpar.maxk-mpar.mink+1),mpar.nk))-1+mpar.mink; %Define asset grid on log-linearspaced

% Meshes of capital and productivity
[meshes.k,  meshes.z] = ndgrid(gri.k,gri.z);

%% 2. Equilibrium in Bewley model with young method
[K,kprime,marginal_k, ~, ~] = K_Agg(0,1,par,mpar,prob.z,meshes,gri); % Solve for equlibrium

%% 3. Plot Policy Functions
figure(1)
plot(gri.k,kprime) % Plot policy functions
hold on
plot(gri.k,gri.k,'k--') % Add 45° line
hold off
title('Policy Value Functions') % Title and legend of the graph
legend({'low productivity','high productivity', '45 degree'},'Location','northwest')
xlabel('assets')
ylabel('consumption')
%% 4. Simulate the economy
% Define an off-grid savings function by interpolation
tic
Saving  = griddedInterpolant({gri.k,gri.z},kprime); % Continuous Savings Function as Linear Interpolant
% Simulate the exogeneous state
PI      = cumsum(prob.z,2);         % Cummulative Transition matrix
epsilon = rand(1,mpar.T);           % Random numbers for simulation
S       = randi(mpar.nz,1,mpar.T);  % Starting value
k       = zeros(1,mpar.T);          % Starting value for NEXT PERIODS ASSETS assets
for t=2:mpar.T
    S(t)    = sum(PI(S(t-1),:)<epsilon(t))+1;   % Update productivity state
    k(t)    = Saving({k(t-1),gri.z(S(t))});     % Update Assest Holdings
end
time(1)     = toc;

% Plot histogram
figure(2)
histogram(k(10001:end),'BinEdges',(gri.k(1:end-1) + gri.k(2:end))/2,'Normalization','probability')
title('Distribution of asset holdings')
hold on

%% 5. Compare solutions
tic % time young method 
[Gamma, StDist] = Young(kprime,gri,mpar,prob.z);
time(2) = toc;

figure(2)
bar(gri.k,marginal_k)
title('Distribution of asset holdings')
legend('Simulation','Direct Calculation')
hold off

disp('Average Asset holdings')
disp('----------------------')
disp(['Simulation: ' num2str(mean(k(10001:end)))])
disp(['Young     : ' num2str(dot(marginal_k',gri.k))]);
disp('----------------------')
disp('Time for calculation')
disp('----------------------')
disp(['Simulation: ' num2str(time(1))])
disp(['Young     : ' num2str(time(2))]);

File: ps2/submit/Sugarkhuu_DynamicMacro_PS2/Ex3-5_julia/Template_Ex_2-5_julia/Ex3.jl

# Specific parameter adjustments
@set! mpar.mink = 0
@set! mpar.maxk = 3 # Maximum assets
@set! par.b = mpar.mink
@set! par.γ = 2
println("-------------------------")
println("Exercise 3: Bewley model")
println("-------------------------")
println(par) # Display economic parameters
## 1. Generate grids, Meshes and Income
@set! gri.k   = collect(range(mpar.mink,mpar.maxk;length=mpar.nk))
# Meshes of capital and productivity
meshes = (
            k = [k for k in gri.k, z in gri.z],
            z = [z for k in gri.k, z in gri.z]
         )

## 2. Equilibrium in Bewley model with Young method
K,kprime,marginal_k, = K_Agg(0,1,par,mpar,Π,meshes,gri) # Solve for equlibrium

## 3. Plot Policy Functions
figure1 = plot(gri.k,kprime,labels=["low productivity" "high productivity"],legend = :topleft) # Plot policy functions
plot!(gri.k,gri.k,linestyle = :dash,linecolor= :black,labels = nothing) # Add 45° line
title!("Policy Value Functions")
xlabel!("assets")
ylabel!("saving")

## 4. Simulate the economy
# Define an off-grid savings function by interpolation
timer = time()
Saving = interpolate((gri.k,gri.z),kprime,Gridded(Linear())) # Continuous Savings Function as Linear Interpolant
# Simulate the exogeneous state
Random.seed!(42)
PI      = cumsum(Π,dims = 2)         # Cummulative Transition matrix
ϵ       = rand(mpar.T)           # Random numbers for simulation
S       = rand(1:mpar.nz,mpar.T)  # Starting value
k       = zeros(mpar.T)          # Starting value for NEXT PERIODS ASSETS assets
for t=2:mpar.T
    S[t]    = count(PI[S[t-1],:].< ϵ[t])+1   # Update productivity state
    k[t]    = Saving(k[t-1],gri.z[S[t]])     # Update Assest Holdings
end
timeSim   = time() - timer

# Plot histogram
figure2 = histogram(k[10001:end],bins=(gri.k[1:end-1] + gri.k[2:end])/2,normalize= :probability,label="Simulation")
title!("Distribution of asset holdings")

## 5. Compare solutions
bar!(gri.k,marginal_k,label="Direct Calculation",legend= :topleft)

timer = time() # time Young method 
Γ, StDist = Young(kprime,gri,mpar,Π)
timeYoung = time() - timer

println("")
println("Average Asset holdings")
println("----------------------")
println("Simulation: ",sum(k[10001:end])/length(k[10001:end]))
println("Young     : ",K)
println("----------------------")
println("Time for calculation")
println("--------------------")
println("Simulation: ",timeSim)
println("Young     : ",timeYoung)